internal/provider/resource_vault_secrets_app.go

```diff
@@ -0,0 +1,168 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	sharedmodels "github.com/hashicorp/hcp-sdk-go/clients/cloud-shared/v1/models"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	clients "github.com/hashicorp/terraform-provider-hcp/internal/clients"
+)
+
+func NewVaultSecretsAppResource() resource.Resource {
+	return &resourceVaultsecretsApp{}
+}
+
+type resourceVaultsecretsApp struct {
+	client *clients.Client
+}
+
+func (r *resourceVaultsecretsApp) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_vault_secrets_app"
+}
+
+func (r *resourceVaultsecretsApp) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "The Vault Secrets app resource manages an application.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:    true,
+				Description: "Required ID field that is set to the app name.",
+			},
+			"app_name": schema.StringAttribute{
+				Required:    true,
+				Description: "The Vault Secrets App name.",
+				Validators: []validator.String{
+					stringvalidator.LengthBetween(3, 36),
+					stringvalidator.RegexMatches(
+						regexp.MustCompile(`^[-\da-zA-Z]{3,36}$`),
+						"must contain only letters, numbers or hyphens",
+					),
+				},
+			},
+			"description": schema.StringAttribute{
+				Description: "The Vault Secrets app name description",
+				Optional:    true,
+			},
+		},
+	}
+}
+
+func (r *resourceVaultsecretsApp) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	client, ok := req.ProviderData.(*clients.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *clients.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+	r.client = client
+}
+
+type App struct {
```

Comment:
This name might be too generic. Would something like `VaultSecretsApp` be better instead?